src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNameParser.cs

```diff
@@ -172,34 +173,56 @@ internal static bool TryParseAnonymousTypeParameterName(string typeParameterName
             return false;
         }
 
+        private const int sha256LengthBytes = 32;
+        private const int sha256LengthHexChars = sha256LengthBytes * 2;
+
         // A full metadata name for a generic file-local type looks like:
         // <ContainingFile>FN__ClassName`A
-        // where 'N' is the syntax tree ordinal, 'A' is the arity,
+        // where 'N' is the SHA256 checksum of the original file path, 'A' is the arity,
         // and 'ClassName' is the source name of the type.
         //
         // The "unmangled" name of a generic file-local type looks like:
         // <ContainingFile>FN__ClassName
-        private static readonly Regex s_fileTypeOrdinalPattern = new Regex(@"<([a-zA-Z_0-9]*)>F(\d)+__", RegexOptions.Compiled);
+        private static readonly Regex s_fileTypeOrdinalPattern = new Regex($@"<([a-zA-Z_0-9]*)>F([0-9A-F]{{{sha256LengthHexChars}}})__", RegexOptions.Compiled);
 
         /// <remarks>
         /// This method will work with either unmangled or mangled type names as input, but it does not remove any arity suffix if present.
         /// </remarks>
-        internal static bool TryParseFileTypeName(string generatedName, [NotNullWhen(true)] out string? displayFileName, out int ordinal, [NotNullWhen(true)] out string? originalTypeName)
+        internal static bool TryParseFileTypeName(string generatedName, [NotNullWhen(true)] out string? displayFileName, [NotNullWhen(true)] out byte[]? checksum, [NotNullWhen(true)] out string? originalTypeName)
```

Comment:
>out byte[]? checksum [](http://example.com/codeflow?start=141&length=20)

It looks like the two callers both convert `checksum` to `ImmutableArray<byte>`. Consider returning that type instead.

Reply:
This file is also included in `Microsoft.CodeAnalysis.CSharp.ResultProvider.csproj`, where `System.Collections.Immutable` is not available.

We could consider moving the function somewhere else or coming up with a compilation symbol which will only be defined if System.Collections.Immutable is available.